Godeps/_workspace/src/github.com/jbenet/go-datastore/flatfs/flatfs.go

```diff
@@ -0,0 +1,241 @@
+// Package flatfs is a Datastore implementation that stores all
+// objects in a two-level directory structure in the local file
+// system, regardless of the hierarchy of the keys.
+package flatfs
+
+import (
+	"encoding/hex"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-datastore"
+	"github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/jbenet/go-datastore/query"
+)
+
+const (
+	extension    = ".data"
+	maxPrefixLen = 16
+)
+
+var (
+	ErrBadPrefixLen = errors.New("bad prefix length")
+)
+
+type Datastore struct {
+	path string
+	// length of the dir splay prefix, in bytes of hex digits
+	hexPrefixLen int
+}
+
+var _ datastore.Datastore = (*Datastore)(nil)
+
+func New(path string, prefixLen int) (*Datastore, error) {
+	if prefixLen <= 0 || prefixLen > maxPrefixLen {
+		return nil, ErrBadPrefixLen
+	}
+	fs := &Datastore{
+		path: path,
+		// convert from binary bytes to bytes of hex encoding
+		hexPrefixLen: prefixLen * hex.EncodedLen(1),
+	}
+	return fs, nil
+}
+
+var padding = strings.Repeat("_", maxPrefixLen*hex.EncodedLen(1))
+
+func (fs *Datastore) encode(key datastore.Key) (dir, file string) {
+	safe := hex.EncodeToString(key.Bytes())
```

Comment:
@jbenet You mean because the prefixes wouldn't necessarily align with the directory sharding done here? I think prefix search would be still doable, it just might have to iterate multiple of these shard directories. It won't perform very well, as it'll need to filter out a ton of non-matches, though.